src/txdb.cpp

```diff
@@ -93,26 +93,32 @@ bool CCoinsViewDB::BatchWrite(CCoinsMap &mapCoins, const uint256 &hashBlock, boo
     size_t count = 0;
     size_t changed = 0;
     assert(!hashBlock.IsNull());
+    bool unfinished_replay{false};
 
     uint256 old_tip = GetBestBlock();
     if (old_tip.IsNull()) {
-        // We may be in the middle of replaying.
+        // We may be in the process of replaying.
         std::vector<uint256> old_heads = GetHeadBlocks();
         if (old_heads.size() == 2) {
             if (old_heads[0] != hashBlock) {
-                LogPrintLevel(BCLog::COINDB, BCLog::Level::Error, "The coins database detected an inconsistent state, likely due to a previous crash or shutdown. You will need to restart bitcoind with the -reindex-chainstate or -reindex configuration option.\n");
```

Comment:
Is it a safe assumption that this cannot happen except an interrupted replay?

Reply:
Currently we abort with an assert in the next line after the log, so the expectation in master is that this shouldn't happen outside of db corruption. I couldn't find any reports on the internet of users that have run into this error / assert, so it probably doesn't happen very often in practice.
But theoretically speaking, if the coins db gets corrupted, I guess anything can happen depending on the specifics of the corruption?

While working on this PR, I initially went with the different approach of adding an explicit `in_replay` flag to `BatchWrite()`, and setting that flag only during the interrupt flush. I switched to the current approach of detecting this implicitly inside BatchWrite() because it seemed more elegant and led to less boilerplate code, but it's still an option.

Reply:
In commit "validation: Make ReplayBlocks interruptible" (26e78f2ee4a325f20c3b3bfaac5be532ab9e7bac)

> I switched to the current approach of detecting this implicitly inside `BatchWrite()` because it seemed more elegant and led to less boilerplate code, but it's still an option.

I would definitely be curious to see the other approach. Not just because it might be better at detecting inconsistent states like Luke is suggesting, but also because it seems like it would make the code more clear.

I think it would be pretty hard to look at the code after this PR (without looking at the PR itself), and understand why `old_heads[0] != hashBlock` implies that this BatchWrite call is finalizing a replay that is currently being interrupted.

I was also confused by the name of the "unfinished_replay" variable below, because I thought "unfinished replay" was just referring to a replay of an previously unfinished flush. I did not realize it was referring specifically to an *interrupted* replay of an unfinished flush.

Also as a possible followup, for future maintainability of this code I think it could be a good idea to remove the replay logic entirely from BatchWrite(), since ReplayBlocks() is the only BatchWrite() caller that should need it. Other BatchWrite calls should be able to assume and assert that GetHeadBlocks() is empty, and this is just a normal write, not a replay.

Reply:
> I would definitely be curious to see the other approach. Not just because it might be better at detecting inconsistent states like Luke is suggesting, but also because it seems like it would make the code more clear.

See https://github.com/mzumsande/bitcoin/commit/db8232d47068bba9ae1bb8135ef456be731e5ca5  - because of the inheritance the bool `interrupted_replay` needed to be added to quite a few spots, which I didn't love.

> I think it would be pretty hard to look at the code after this PR (without looking at the PR itself), and understand why old_heads[0] != hashBlock implies that this BatchWrite call is finalizing a replay that is currently being interrupted.

I'll add more documentation (if I don't switch to the other approach).

> I was also confused by the name of the "unfinished_replay" variable below, 

Maybe `interrupted_replay` is better? But the issue remains that it's the replay that is being interrupted (or unfinished), and not the flush.

> Also as a possible followup, for future maintainability of this code I think it could be a good idea to remove the replay logic entirely from BatchWrite(), since ReplayBlocks() is the only BatchWrite() caller that should need it. Other BatchWrite calls should be able to assume and assert that GetHeadBlocks() is empty, and this is just a normal write, not a replay.

Seems reasonable for a follow-up, though I haven't looked into details yet.